Split global_time_series task and improve performance

forsyth2 opened the pull request:
## Summary

Corresponding `zppy` PR: https://github.com/E3SM-Project/zppy/pull/722

Objectives:
- Separate out the Classic PDF and Viewer implementations for `global_time_series`. These two output formats deviate substantially and so separating them can help clean up the code.
- Improve performance

Issue resolution:
- Closes #23
- Ideally also closes #21 (by replacing #22)
- Could potentially implement/replace #19 as well

Select one: This pull request is...
- [ ] a bug fix: increment the patch version
- [ ] a small improvement: increment the minor version
- [x] a new feature: increment the minor version
- [ ] an incompatible (non-backwards compatible) API change: increment the major version

## Big Change

- [x] To merge, I will use "Create a merge commit". That is, this change is large enough to require multiple units of work (i.e., it should be multiple commits).

### 1. Does this do what we want it to do?

Required:
- [x] Product Management: I have confirmed with the stakeholders that the objectives above are correct and complete.
- [x] Testing: I have considered likely and/or severe edge cases and have included them in testing.

### 2. Are the implementation details accurate & efficient?

Required:
- [x] Logic: I have visually inspected the entire pull request myself.
- [x] Logic: I have left GitHub comments highlighting important pieces of code logic. I have had these code blocks reviewed by at least one other team member.

If applicable:
- [ ] Dependencies: This pull request introduces a new dependency. I have discussed this requirement with at least one other team member. The dependency is noted in `zppy-interfaces/conda`, not just an `import` statement.

### 3. Is this well documented?

Required:
- [ ] Documentation: by looking at the docs, a new user could easily understand the functionality introduced by this pull request.

### 4. Is this code clean?

Required:
- [x] Readability: The code is as simple as possible and well-commented, such that a new team member could understand what's happening.
- [x] Pre-commit checks: All the pre-commits checks have passed.

If applicable:
- [x] Software architecture: I have discussed relevant trade-offs in design decisions with at least one other team member. It is unlikely that this pull request will increase tech debt.

forsyth2 commented:
The Classic PDF and Viewer modes are now mostly separated. I also added legacy `cfg` tests. I also have subtasks working for `global_time_series`, which has been a great help to testing multiple cases simultaneously. I cleaned up the plotting code for viewers and I made updates to load `rgn` data at once per variable (similar to #19). Not re-loading the same data brought a test on a subset of land vars down from 124 seconds to 93 seconds. 

I'm currently testing the true all-land-variable case, where we use NCO to get every land variable we can, and then go through the `csv` of land vars.

Remaining TODO as of 6/6:
- [ ] It's contradictory to have both `plots_original` and any of the `plots_{component}` parameters set. In particular, since `plots_original` defaults to a list with dependencies on `ocn` data, it will by default cause Viewer mode land plots to have dependencies they don't need. So, this should be fixed. This also segues into a larger design decision below.
- [ ] Check that all land variables from the `csv` (that are available) can be plotted using `plots_lnd="all"` (currently running this `cfg`)
- [ ] Run the other weekly `cfg`s, including the new legacy ones.
- [ ] Run the integration test suite on the output of those `cfg`s.
- [ ] Send to code review

Major design decisions to make/confirm:
1. How to properly handle parameters in the new 2-mode system:
```
        # TODO: it may make sense in the future to refactor this class. Options:
        # 1. Have this be the base class, for paramters used by both output types
        # and then have distinct ClassicParameters and ViewerParameters subclasses.
        # One potential issue is that the type checker will get confused if a
        # common utility function returns Parameters when we know in context
        # it's returning a specific subtype.
        # 2. Have 3 distinct classes: CoreParameters, ClassicParameters, ViewerParameters.
        # The downside here is that there will need to be two variables to hold the parameters:
        # core_parameters and either classic_parameters or viewer_parameters.
        # 3. The simplest thing is just to add input validation --
        # e.g., if in viewer mode, make sure these parameters are set and
        # if in classic mode, make sure these other parameters are set.
        #
        # This would be useful if we want to split `zi-global-time-series` into 2 distinct entry points,
        # each with specialized parameters.
```
2. Per the last point there, do we want the 2 modes to have separate entry points or for the mode split to happen after the entry point to `zppy-interfaces`?
3. Will the Classic PDF _always_ be in the 4x2 grid? That is, do we need user-specified `nrows` and `ncols` still?
4. The Classic PDF will _never_ be used for plots other than the original 8, correct?
5. We _never_ want a viewer for the original 8 plots, correct?

forsyth2 commented:
> How to properly handle parameters in the new 2-mode system:

To further expand, there are several points in the workflow where we can separate into the 2 modes. From least to most change:

1. _Current implementation_ -- one `zppy` `[global_time_series]` task, one `zppy-interfaces` entry point. Classic PDF vs Viewer mode is determined based on `make_viewer` parameter. Because of the combined entry point, some parameters passed in are superfluous.
2. Separate `zppy-interfaces` entry points. That is, `global_time_series.bash` in `zppy` would have two distinct `zppy-interfaces` calls -- one for the case of Classic PDF and one for the case of Viewer. Each entry point could have specialized parameters.
3. Distinct tasks in the `zppy` `cfg`, e.g., `[global_time_series_classic]` and `[global_time_series_viewer]` tasks. This level of change would almost certainly be a breaking change (i.e., not-backward-compatible, increment major version number). That is, at this level, the user has two distinct tasks they want to run.

(Please note that the new subtask functionality allows both a Classic PDF job and a Viewer job to run simultaneously.)

chengzhuzhang commented:
@forsyth2 thanks for the PR, I have some initial review with questions:
Since the update is tested with true all-land-variable case, as well as including the classic coupled figures, how the performance look like?
The new scripts under /viewer seem to have some duplicated code for component wise anlysis and plotting, can use some refactoring to make the code more maintainable. 
For refactor at this scale, it would be helpful to have @tomvothecoder to also have a review..

forsyth2 commented:
> Since the update is tested with true all-land-variable case, as well as including the classic coupled figures, how the performance look like?

I put some initial performance numbers in https://github.com/E3SM-Project/zppy/pull/722#issuecomment-2957217188. 

For this latest run, we find:
```bash
cd /lcrc/group/e3sm/ac.forsyth2/zppy_weekly_comprehensive_v3_output/test-zppy-gts-split-20250610-try3/v3.LR.historical_0051/post/scripts/ && grep -H "Elapsed time" global_time_series_*.o*
# global_time_series_all_lnd_var_viewer_1985-1995.o759913:Elapsed time: 5526 seconds
# global_time_series_classic_original_8_1985-1995.o759915:Elapsed time: 65 seconds
# global_time_series_classic_original_8_no_ocn_1985-1995.o759914:Elapsed time: 63 seconds
# global_time_series_specific_var_viewers_1985-1995.o759912:Elapsed time: 79 seconds

# We can see here that the all_lnd var case took 5526 seconds = 92.1 minutes = 1.5 hours
# While we're here, let's also check how long it takes NCO to process all the lnd variables:
grep -H "Elapsed time" ts_lnd_monthly_glb_*.o*
# ts_lnd_monthly_glb_1985-1989-0005.o759899:Elapsed time 1m26s
# ts_lnd_monthly_glb_1985-1989-0005.o759899:Elapsed time: 91 seconds
# ts_lnd_monthly_glb_1990-1994-0005.o759900:Elapsed time 1m25s
# ts_lnd_monthly_glb_1990-1994-0005.o759900:Elapsed time: 90 seconds

# None of the other tests have the subtasks, so they're only running the Classic PDF mode.

cd /lcrc/group/e3sm/ac.forsyth2/zppy_weekly_comprehensive_v2_output/test-zppy-gts-split-20250610-try3/v2.LR.historical_0201/post/scripts/ && grep -H "Elapsed time" global_time_series_*.o*
# global_time_series_1980-1990.o759882:Elapsed time: 92 seconds

cd /lcrc/group/e3sm/ac.forsyth2/zppy_weekly_legacy_3.0.0_comprehensive_v3_output/test-zppy-gts-split-20250610-try3/v3.LR.historical_0051/post/scripts/ && grep -H "Elapsed time" global_time_series_*.o*
# global_time_series_1985-1995.o759978:Elapsed time: 89 seconds

cd /lcrc/group/e3sm/ac.forsyth2/zppy_weekly_legacy_3.0.0_comprehensive_v2_output/test-zppy-gts-split-20250610-try3/v2.LR.historical_0201/post/scripts/ && grep -H "Elapsed time" global_time_series_*.o*
# global_time_series_1980-1990.o759948:Elapsed time: 85 seconds
```

> The new scripts under /viewer seem to have some duplicated code for component wise anlysis and plotting, can use some refactoring to make the code more maintainable.

Can you point to a specific code block (i.e., comment on it in the PR)? You mean code is duplicated amongst components in Viewer mode, correct? (i.e., not that code is duplicated between Viewer and Classic PDF modes). I'm looking at the `/viewer` subdir and it seems pretty abstracted to me. E.g., `construct_generic_variables` and `set_var` are used by multiple components.

> For refactor at this scale, it would be helpful to have @tomvothecoder to also have a review..

Sure, @tomvothecoder please review if you have the bandwidth.

@chengzhuzhang in the meantime, can you review https://github.com/E3SM-Project/zppy/pull/720 so we can at least get that part merged? (The first two commits of that PR are shared with https://github.com/E3SM-Project/zppy/pull/722.)

forsyth2 commented on zppy_interfaces/global_time_series/viewer/coupled_global_plotting.py:
```diff
@@ -0,0 +1,93 @@
+            traceback.print_exc()
+            logger.error(f"plot_generic failed. Invalid plot={plot_name}, rgn={rgn}")
+            invalid_plots.append(plot_name)
+        i += 1
```
Yes, it appears this didn't cause any issues because `i` gets rewritten on each iteration, but yes it should be removed. Probably a holdover from the original combined plotting.

forsyth2 commented:
It looks like that build error also occurs on the last build I have for a different PR: https://github.com/E3SM-Project/zppy-interfaces/actions/runs/15645218263/job/44081356964?pr=25, so it's definitely an issue on the `main` branch.

forsyth2 commented:
Oh I see, by comparing to `zppy`'s commit history, it looks like we need the equivalent of https://github.com/E3SM-Project/zppy/pull/715/files

forsyth2 commented:
Ok it looks like https://github.com/E3SM-Project/zppy-interfaces/pull/28/files fixed the problem. You can ignore above tags @tomvothecoder & @xylar.

forsyth2 commented:
@chengzhuzhang Ok, this pull request (and it's accompanying `zppy` PR https://github.com/E3SM-Project/zppy/pull/722) are ready for re-review. Specifically, here's what changed:
- `zppy-interfaces` side: https://github.com/E3SM-Project/zppy-interfaces/pull/26/commits/1ca04deaa9a81d91f64e1bd55ed2dd490ad26af8
- `zppy` side: https://github.com/E3SM-Project/zppy/pull/722/commits/a3013404e36571963f59c6d69dc58c6ac675db15

I explain in detail the reason for the redesign in https://github.com/E3SM-Project/zppy/pull/722#discussion_r2157910395, but the high-level summary is this: The original design called for a _perfect_ separation of Classic PDF and Viewer modes. That is, original plots get plotted in a PDF and component plots get plotted in the Viewer. The backwards compatibility considerations contradicted that assumption of perfect separation, requiring a redesign.

chengzhuzhang commented:
@forsyth2 I'm trying to re-review the PRs. I'm looking at results here, https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v3_www/test-gts-split-20250701/v3.LR.historical_0051/global_time_series/
I'm wondering if I can find the .cfg files for each case, and is particularly interested in this case: [global_time_series_viewer_both_1985-1995_results/](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v3_www/test-gts-split-20250701/v3.LR.historical_0051/global_time_series/global_time_series_viewer_both_1985-1995_results/), which I would think should be the case to present in the example cfg file for the user. In the viewer for [original plots](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v3_www/test-gts-split-20250701/v3.LR.historical_0051/global_time_series/global_time_series_viewer_both_1985-1995_results/table_original/index.html), it has two rows, one for PDF, the other for PNG, and it looks like the PDF links are all broken. Not sure why we need both here...

forsyth2 commented:
> I'm wondering if I can find the .cfg files for each case

https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v3_www/test-gts-split-20250701/v3.LR.historical_0051/global_time_series/ is generated with just one `.cfg`:
```
/lcrc/group/e3sm/public_html/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v3_www/test-gts-split-20250701/v3.LR.historical_0051/provenance.20250701_183021_093072.cfg
```
This can be viewed on the web at https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v3_www/test-gts-split-20250701/v3.LR.historical_0051/provenance.20250701_183021_093072.cfg.

<details>

<summary> Full cfg </summary>

```
[default]
case = "v3.LR.historical_0051"
constraint = ""
dry_run = "False"
environment_commands = ""
fail_on_dependency_skip = True
infer_path_parameters = False
infer_section_parameters = False
input = /lcrc/group/e3sm2/ac.wlin//E3SMv3/v3.LR.historical_0051
input_subdir = archive/atm/hist
mapping_file = "map_ne30pg2_to_cmip6_180x360_aave.20200201.nc"
output = "/lcrc/group/e3sm/ac.forsyth2/zppy_weekly_comprehensive_v3_output/test-gts-split-20250701/v3.LR.historical_0051"
partition = "debug"
qos = "regular"
www = "/lcrc/group/e3sm/public_html/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v3_www/test-gts-split-20250701"
years = "1985:1989:2",

[climo]
active = True
walltime = "00:30:00"

  [[ atm_monthly_180x360_aave ]]
  frequency = "monthly"
  input_files = "eam.h0"
  input_subdir = "archive/atm/hist"
  vars = ""

  [[ atm_monthly_diurnal_8xdaily_180x360_aave ]]
  frequency = "diurnal_8xdaily"
  input_files = "eam.h3"
  input_subdir = "archive/atm/hist"
  vars = "PRECT"

  [[ land_monthly_climo ]]
  frequency = "monthly"
  input_files = "elm.h0"
  input_subdir = "archive/lnd/hist"
  mapping_file = "map_r05_to_cmip6_180x360_aave.20231110.nc"
  vars = ""

[ts]
active = True
walltime = "00:30:00"

  [[ atm_monthly_180x360_aave ]]
  frequency = "monthly"
  input_files = "eam.h0"
  input_subdir = "archive/atm/hist"

  [[ atm_daily_180x360_aave ]]
  frequency = "daily"
  input_files = "eam.h1"
  input_subdir = "archive/atm/hist"
  vars = "PRECT"

  [[ rof_monthly ]]
  extra_vars = 'areatotal2'
  frequency = "monthly"
  input_files = "mosart.h0"
  input_subdir = "archive/rof/hist"
  mapping_file = ""
  vars = "RIVER_DISCHARGE_OVER_LAND_LIQ"

  [[ atm_monthly_glb ]]
  # Note global average won't work for 3D variables.
  frequency = "monthly"
  input_files = "eam.h0"
  input_subdir = "archive/atm/hist"
  mapping_file = "glb"
  years = "1985:1995:5",

  [[ lnd_monthly_glb ]]
  frequency = "monthly"
  input_files = "elm.h0"
  input_subdir = "archive/lnd/hist"
  mapping_file = "glb"
  vars = ""
  years = "1985:1995:5",

  [[ land_monthly ]]
  extra_vars = "landfrac"
  frequency = "monthly"
  input_files = "elm.h0"
  input_subdir = "archive/lnd/hist"
  mapping_file = "map_r05_to_cmip6_180x360_aave.20231110.nc"
  vars = "FSH,RH2M,LAISHA,LAISUN,QINTR,QOVER,QRUNOFF,QSOIL,QVEGE,QVEGT,SOILICE,SOILLIQ,SOILWATER_10CM,TSA,TSOI,H2OSNO,TOTLITC,CWDC,SOIL1C,SOIL2C,SOIL3C,SOIL4C,WOOD_HARVESTC,TOTVEGC,NBP,GPP,AR,HR"

[e3sm_to_cmip]
active = True
frequency = "monthly"
ts_num_years=2
walltime = "00:30:00"

  [[ atm_monthly_180x360_aave ]]
  input_files = "eam.h0"
  ts_subsection = "atm_monthly_180x360_aave"

  [[ land_monthly ]]
  input_files = "elm.h0"
  ts_subsection = "land_monthly"

# TODO: Add "tc_analysis" back in after empty dat is resolved.
# [tc_analysis]
# active = True
# walltime = "00:30:00"

[e3sm_diags]
active = True
climo_diurnal_frequency = "diurnal_8xdaily"
climo_diurnal_subsection = "atm_monthly_diurnal_8xdaily_180x360_aave"
climo_subsection = "atm_monthly_180x360_aave"
environment_commands = "source /lcrc/soft/climate/e3sm-unified/load_latest_e3sm_unified_chrysalis.sh"
grid = '180x360_aave'
multiprocessing = True
num_workers = 8
partition = "compute"
qos = "regular"
ref_final_yr = 1986
ref_start_yr = 1985
ref_years = "1985-1986",
# Include all sets
# min_case_e3sm_diags_depend_on_climo: "lat_lon","zonal_mean_xy","zonal_mean_2d","polar","cosp_histogram","meridional_mean_2d","annual_cycle_zonal_mean","zonal_mean_2d_stratosphere","aerosol_aeronet","aerosol_budget",
# min_case_e3sm_diags_depend_on_ts: "enso_diags","qbo",
# min_case_e3sm_diags_diurnal_cycle: "diurnal_cycle",
# min_case_e3sm_diags_streamflow: "streamflow",
# min_case_e3sm_diags_tc_analysis: "tc_analysis",
# min_case_e3sm_diags_tropical_subseasonal: "tropical_subseasonal",
# TODO: Add "tc_analysis" back in after empty dat is resolved.
# TODO: Add "aerosol_budget" back in once that's working for v3.
sets = "lat_lon","zonal_mean_xy","zonal_mean_2d","polar","cosp_histogram","meridional_mean_2d","annual_cycle_zonal_mean","zonal_mean_2d_stratosphere","enso_diags","qbo","diurnal_cycle","streamflow","tropical_subseasonal","aerosol_aeronet",
short_name = "v3.LR.historical_0051"
ts_daily_subsection = "atm_daily_180x360_aave"
ts_num_years = 2
ts_subsection = "atm_monthly_180x360_aave"
walltime = "5:00:00"
years = "1987:1989:2"
# Reference paths
# Used for mvo and mvm, if ts_num_years is set
obs_ts = "/lcrc/group/e3sm/diagnostics/observations/Atm/time-series/"
# mvo & mvm tc_analysis only
tc_obs = "/lcrc/group/e3sm/diagnostics/observations/Atm/tc-analysis/"

  [[ atm_monthly_180x360_aave ]]
  # Reference paths
  reference_data_path = "/lcrc/group/e3sm/diagnostics/observations/Atm/climatology/"
  # mvo diurnal_cycle only
  # NOTE: This is NOT the inference zppy would have made!
  dc_obs_climo = '/lcrc/group/e3sm/public_html/e3sm_diags_test_data/unit_test_complete_run/obs/climatology'
  # mvo streamflow only
  streamflow_obs_ts = "/lcrc/group/e3sm/diagnostics/observations/Atm/time-series/"

  [[ atm_monthly_180x360_aave_mvm ]]
  # Test model-vs-model using the same files as the reference
  diff_title = "Difference"
  partition = "compute"
  qos = "regular"
  ref_name = "v3.LR.historical_0051"
  run_type = "model_vs_model"
  short_ref_name = "same simulation"
  swap_test_ref = False
  tag = "model_vs_model"
  ts_num_years_ref = 2
  # Reference paths
  reference_data_path = "/lcrc/group/e3sm/ac.forsyth2/zppy_weekly_comprehensive_v3_output/test-gts-split-20250701/v3.LR.historical_0051/post/atm/180x360_aave/clim"
  # mvm streamflow only
  gauges_path = "/lcrc/group/e3sm/diagnostics/observations/Atm/time-series/GSIM/GSIM_catchment_characteristics_all_1km2.csv"
  reference_data_path_ts_rof = "/lcrc/group/e3sm/ac.forsyth2/zppy_weekly_comprehensive_v3_output/test-gts-split-20250701/v3.LR.historical_0051/post/rof/native/ts/monthly"
  # mvm diurnal_cycle only
  reference_data_path_climo_diurnal = "/lcrc/group/e3sm/ac.forsyth2/zppy_weekly_comprehensive_v3_output/test-gts-split-20250701/v3.LR.historical_0051/post/atm/180x360_aave/clim_diurnal_8xdaily"
  # mvm "enso_diags", "qbo", "area_mean_time_series"
  reference_data_path_ts = "/lcrc/group/e3sm/ac.forsyth2/zppy_weekly_comprehensive_v3_output/test-gts-split-20250701/v3.LR.historical_0051/post/atm/180x360_aave/ts/monthly"
  # mvm tropical_subseasonal only
  reference_data_path_ts_daily = "/lcrc/group/e3sm/ac.forsyth2/zppy_weekly_comprehensive_v3_output/test-gts-split-20250701/v3.LR.historical_0051/post/atm/180x360_aave/ts/daily"

  [[ lnd_monthly_mvm_lnd ]]
  # Test model-vs-model using the same files as the reference
  climo_subsection = "land_monthly_climo"
  diff_title = "Difference"
  partition = "compute"
  qos = "regular"
  ref_name = "v3.LR.historical_0051"
  run_type = "model_vs_model"
  sets = "lat_lon_land",
  short_ref_name = "same simulation"
  swap_test_ref = False
  tag = "model_vs_model"
  ts_num_years_ref = 2
  # Reference paths
  reference_data_path = "/lcrc/group/e3sm/ac.forsyth2/zppy_weekly_comprehensive_v3_output/test-gts-split-20250701/v3.LR.historical_0051/post/lnd/180x360_aave/clim"


[mpas_analysis]
active = True
anomalyRefYear = 1985
climo_years = "1985-1989", "1990-1995",
enso_years = "1985-1989", "1990-1995",
mesh = "IcoswISC30E3r5"
parallelTaskCount = 6
partition = "compute"
qos = "regular"
shortTermArchive = True
ts_years = "1985-1989", "1985-1995",
walltime = "02:00:00"

[global_time_series]
active = True
climo_years = "1985-1989", "1990-1995",
environment_commands = "source /gpfs/fs1/home/ac.forsyth2/miniforge3/etc/profile.d/conda.sh; conda activate zi-gts-split-20250603"
experiment_name = "v3.LR.historical_0051"
figstr = "v3.LR.historical_0051"
#moc_file=mocTimeSeries_1985-1995.nc
# plots_lnd = "FSH,RH2M,LAISHA,LAISUN,QINTR,QOVER,QRUNOFF,QSOIL,QVEGE,QVEGT,SOILWATER_10CM,TSA,H2OSNO,TOTLITC,CWDC,SOIL1C,SOIL2C,SOIL3C,SOIL4C,WOOD_HARVESTC,TOTVEGC,NBP,GPP,AR,HR"
ts_num_years = 5
ts_years = "1985-1989", "1985-1995",
walltime = "00:30:00"
years = "1985-1995",

  # Important parameter combinations ##########################################

  [[ viewer_both ]]
  # 1. make_viewer = True, plots_original set, >= 1 plots_<component> set
  # NOTE: This is the case displayed in examples/post.v3.LR.historical_zppy_v3.cfg
  make_viewer = True
  moc_file = "mocTimeSeries_1985-1995.nc"
  plots_atm = "TREFHT"
  plots_lnd = "FSH,RH2M,LAISHA,LAISUN,QINTR,QOVER,QRUNOFF,QSOIL,QVEGE,QVEGT,SOILWATER_10CM,TSA,H2OSNO,TOTLITC,CWDC,SOIL1C,SOIL2C,SOIL3C,SOIL4C,WOOD_HARVESTC,TOTVEGC,NBP,GPP,AR,HR"
  # Default plots_original = "net_toa_flux_restom,global_surface_air_temperature,toa_radiation,net_atm_energy_imbalance,change_ohc,max_moc,change_sea_level,net_atm_water_imbalance",

  [[ viewer_original ]]
  # 2. make_viewer = True, plots_original set, 0 plots_<component> set
  make_viewer = True
  moc_file = "mocTimeSeries_1985-1995.nc"
  # Default plots_original = "net_toa_flux_restom,global_surface_air_temperature,toa_radiation,net_atm_energy_imbalance,change_ohc,max_moc,change_sea_level,net_atm_water_imbalance",

  [[ viewer_component ]]
  # 3. make_viewer = True, plots_original set, >= 1 plots_<component> set
  make_viewer = True
  plots_atm = "TREFHT"
  plots_lnd = "FSH,RH2M,LAISHA,LAISUN,QINTR,QOVER,QRUNOFF,QSOIL,QVEGE,QVEGT,SOILWATER_10CM,TSA,H2OSNO,TOTLITC,CWDC,SOIL1C,SOIL2C,SOIL3C,SOIL4C,WOOD_HARVESTC,TOTVEGC,NBP,GPP,AR,HR"
  plots_original = ""

  # We can ignore this case for image checking.
  # [[ viewer_none ]]
  # # 4. make_viewer = True, plots_original not set, 0 plots_<component> set
  # make_viewer = True
  # plots_original = ""

  [[ classic_pdf_both ]]
  # 5. make_viewer = False, plots_original set, >= 1 plots_<component> set
  # Default make_viewer = False
  moc_file = "mocTimeSeries_1985-1995.nc"
  plots_atm = "TREFHT"
  plots_lnd = "FSH,RH2M,LAISHA,LAISUN,QINTR,QOVER,QRUNOFF,QSOIL,QVEGE,QVEGT,SOILWATER_10CM,TSA,H2OSNO,TOTLITC,CWDC,SOIL1C,SOIL2C,SOIL3C,SOIL4C,WOOD_HARVESTC,TOTVEGC,NBP,GPP,AR,HR"
  # Default plots_original = "net_toa_flux_restom,global_surface_air_temperature,toa_radiation,net_atm_energy_imbalance,change_ohc,max_moc,change_sea_level,net_atm_water_imbalance",

  [[ classic_pdf_original ]]
  # 6. make_viewer = False, plots_original set, 0 plots_<component> set
  # NOTE: This is the default case -- where all of the above parameters are at their default values.
  # Default make_viewer = False
  moc_file = "mocTimeSeries_1985-1995.nc"
  # Default plots_original = "net_toa_flux_restom,global_surface_air_temperature,toa_radiation,net_atm_energy_imbalance,change_ohc,max_moc,change_sea_level,net_atm_water_imbalance",

  [[ classic_pdf_component ]]
  # 7. make_viewer = False, plots_original set, >= 1 plots_<component> set
  # Default make_viewer = False
  plots_atm = "TREFHT"
  plots_lnd = "FSH,RH2M,LAISHA,LAISUN,QINTR,QOVER,QRUNOFF,QSOIL,QVEGE,QVEGT,SOILWATER_10CM,TSA,H2OSNO,TOTLITC,CWDC,SOIL1C,SOIL2C,SOIL3C,SOIL4C,WOOD_HARVESTC,TOTVEGC,NBP,GPP,AR,HR"
  plots_original = ""

  # We can ignore this case for image checking.
  # [[ classic_pdf_none ]]
  # # 8. make_viewer = False, plots_original not set, 0 plots_<component> set
  # # Default make_viewer = False
  # plots_original = ""

  # Special cases #############################################################

  [[ all_lnd_var_viewer ]]
  # 1. plot ALL land variables
  make_viewer = True
  partition = "compute"
  plots_lnd = "all"
  plots_original = ""
  walltime = "03:00:00"

  [[ classic_original_8_no_ocn ]]
  # 2. exclude ocean from original plots
  plots_original = "net_toa_flux_restom,global_surface_air_temperature,toa_radiation,net_atm_energy_imbalance,net_atm_water_imbalance"


[ilamb]
active = True
e3sm_to_cmip_atm_subsection = "atm_monthly_180x360_aave"
e3sm_to_cmip_land_subsection = "land_monthly"
ilamb_obs = "/lcrc/group/e3sm/diagnostics/ilamb_data"
nodes = 8
partition = "debug"
short_name = "v3.LR.historical_0051"
ts_atm_subsection = "atm_monthly_180x360_aave"
ts_land_subsection = "land_monthly"
ts_num_years = 2
walltime = "2:00:00"
years = "1985:1989:4"
```

</details>

For the https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v3_www/test-gts-split-20250701/v3.LR.historical_0051/global_time_series/global_time_series_viewer_both_1985-1995_results/ case in particular, that's directed by this section:

```
  [[ viewer_both ]]
  # 1. make_viewer = True, plots_original set, >= 1 plots_<component> set
  # NOTE: This is the case displayed in examples/post.v3.LR.historical_zppy_v3.cfg
  make_viewer = True
  moc_file = "mocTimeSeries_1985-1995.nc"
  plots_atm = "TREFHT"
  plots_lnd = "FSH,RH2M,LAISHA,LAISUN,QINTR,QOVER,QRUNOFF,QSOIL,QVEGE,QVEGT,SOILWATER_10CM,TSA,H2OSNO,TOTLITC,CWDC,SOIL1C,SOIL2C,SOIL3C,SOIL4C,WOOD_HARVESTC,TOTVEGC,NBP,GPP,AR,HR"
  # Default plots_original = "net_toa_flux_restom,global_surface_air_temperature,toa_radiation,net_atm_energy_imbalance,change_ohc,max_moc,change_sea_level,net_atm_water_imbalance",
```

> it has two rows, one for PDF, the other for PNG,

Yes, correct. I went with that design because using the example `cfg`, we had:
```
> cd /lcrc/group/e3sm/public_html/diagnostic_output/ac.forsyth2/zppy_example_v3_20250617_try3/v3.LR.historical_0051/global_time_series/global_time_series_1985-2014_results
> ls | grep original
v3.LR.historical_0051_glb_original.pdf
v3.LR.historical_0051_glb_original.png
v3.LR.historical_0051_n_original.pdf
v3.LR.historical_0051_n_original.png
v3.LR.historical_0051_s_original.pdf
v3.LR.historical_0051_s_original.png
```
So, all the pdfs and png's are there, but they aren't linked to in the viewer, so there's no direct path for the user to click to. From https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_example_v3_20250617_try3/v3.LR.historical_0051/global_time_series/global_time_series_1985-2014_results/, they can click on "Original" in the viewer, which gives them exactly one of those 6: [glb_original, PDF version](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_example_v3_20250617_try3/v3.LR.historical_0051/global_time_series/global_time_series_1985-2014_results/v3.LR.historical_0051_glb_original.pdf)

With this PR, all 6 that are produced actually get linked to.

> it looks like the PDF links are all broken

I thought I wrote a comment about this, but it looks like I only made inline notes:
```
def create_viewer_for_original(parameters: Parameters) -> str:
    component: str = "original"
    logger.info(f"Creating viewer for {component}")
    index_name = f"zppy global time-series plot: {parameters.experiment_name} {component} component ({parameters.year1}-{parameters.year2})"
    viewer = OutputViewer(path=parameters.results_dir, index_name=index_name)
    viewer.add_page(f"table_{component}", parameters.regions)
    viewer.add_group("Original Plots")
    # PDFs -- these don't show up when clicked on, but they CAN be downloaded
    row_title: str = "Original Plots, PDFs"
    viewer.add_row(row_title)
    for rgn in parameters.regions:
        viewer.add_col(
            f"{parameters.figstr}_{rgn}_{component}.pdf",
            is_file=True,
            title=f"{rgn}_{component}",
        )
    # PNGs -- these show up when clicked on
    row_title = "Original Plots, PNGs"
    viewer.add_row(row_title)
    for rgn in parameters.regions:
        viewer.add_col(
            f"{parameters.figstr}_{rgn}_{component}.png",
            is_file=True,
            title=f"{rgn}_{component}",
        )
    url = viewer.generate_page()
    viewer.generate_viewer()
    return url
```
The download functionality works; they are simply not displayed in-page. I don't think the underlying viewer functionality is built to display PDFs rather than PNGs.

> Not sure why we need both here 

I kept both for backwards compatibility, as described above. Personally, I'd be fine only displaying the PNG row since 1) that works and 2) it matches how the other plots are displayed (also in PNG form).

forsyth2 commented:
@chengzhuzhang To run standalone, you would set those parameters up according to the command in [global_time_series.bash](https://github.com/E3SM-Project/zppy/blob/issue-23-gts-split/zppy/templates/global_time_series.bash):
```
zi-global-time-series --subsection ${subsection} --use_ocn {{ use_ocn }} --input {{ input }} --input_subdir {{ input_subdir }} --moc_file {{ moc_file }} --case_dir {{ output }} --experiment_name {{ experiment_name }} --figstr {{ figstr }} --color {{ color }} --ts_num_years {{ ts_num_years }} --plots_original {{ plots_original }} --plots_atm {{ plots_atm }} --plots_ice {{ plots_ice }} --plots_lnd {{ plots_lnd }} --plots_ocn {{ plots_ocn }} --nrows {{ nrows }} --ncols {{ ncols }} --results_dir ${results_dir} --regions {{ regions }} --make_viewer {{ make_viewer }} --start_yr {{ year1 }} --end_yr {{ year2 }}
```
filled in with the parameters of the test cases noted in https://github.com/E3SM-Project/zppy/pull/722#issuecomment-3247295197 (under "Full list of test cases")

For example, on my latest run, that line gets realized as:

| Test case | Key points | zppy-interfaces call |
| --- | --- | --- |
| Viewers for both original & component | `plots_original`, `plots_atm`, `plots_lnd`, `make_viewer` are set. | `zi-global-time-series --subsection ${subsection} --use_ocn True --input /lcrc/group/e3sm2/ac.wlin//E3SMv3/v3.LR.historical_0051 --input_subdir archive/ocn/hist --moc_file mocTimeSeries_1985-1995.nc --case_dir /lcrc/group/e3sm/ac.forsyth2/zppy_weekly_comprehensive_v3_output/test_issue-23-rebased-20250903/v3.LR.historical_0051 --experiment_name v3.LR.historical_0051 --figstr v3.LR.historical_0051 --color Blue --ts_num_years 5 --plots_original net_toa_flux_restom,global_surface_air_temperature,toa_radiation,net_atm_energy_imbalance,change_ohc,max_moc,change_sea_level,net_atm_water_imbalance --plots_atm TREFHT --plots_ice None --plots_lnd FSH,RH2M,LAISHA,LAISUN,QINTR,QOVER,QRUNOFF,QSOIL,QVEGE,QVEGT,SOILWATER_10CM,TSA,H2OSNO,TOTLITC,CWDC,SOIL1C,SOIL2C,SOIL3C,SOIL4C,WOOD_HARVESTC,TOTVEGC,NBP,GPP,AR,HR --plots_ocn None --nrows 4 --ncols 2 --results_dir ${results_dir} --regions glb,n,s --make_viewer True --start_yr 1985 --end_yr 1995` |
| Viewer for just original | `plots_original`, `make_viewer` are set; `plots_atm`, `plots_lnd` are not. | `zi-global-time-series --subsection ${subsection} --use_ocn True --input /lcrc/group/e3sm2/ac.wlin//E3SMv3/v3.LR.historical_0051 --input_subdir archive/ocn/hist --moc_file mocTimeSeries_1985-1995.nc --case_dir /lcrc/group/e3sm/ac.forsyth2/zppy_weekly_comprehensive_v3_output/test_issue-23-rebased-20250903/v3.LR.historical_0051 --experiment_name v3.LR.historical_0051 --figstr v3.LR.historical_0051 --color Blue --ts_num_years 5 --plots_original net_toa_flux_restom,global_surface_air_temperature,toa_radiation,net_atm_energy_imbalance,change_ohc,max_moc,change_sea_level,net_atm_water_imbalance --plots_atm None --plots_ice None --plots_lnd None --plots_ocn None --nrows 4 --ncols 2 --results_dir ${results_dir} --regions glb,n,s --make_viewer True --start_yr 1985 --end_yr 1995` |
| Viewer for just component | `plots_atm`, `plots_lnd`, `make_viewer` are set; `plots_original` is not. | `zi-global-time-series --subsection ${subsection} --use_ocn False --input /lcrc/group/e3sm2/ac.wlin//E3SMv3/v3.LR.historical_0051 --input_subdir archive/ocn/hist --moc_file None --case_dir /lcrc/group/e3sm/ac.forsyth2/zppy_weekly_comprehensive_v3_output/test_issue-23-rebased-20250903/v3.LR.historical_0051 --experiment_name v3.LR.historical_0051 --figstr v3.LR.historical_0051 --color Blue --ts_num_years 5 --plots_original None --plots_atm TREFHT --plots_ice None --plots_lnd FSH,RH2M,LAISHA,LAISUN,QINTR,QOVER,QRUNOFF,QSOIL,QVEGE,QVEGT,SOILWATER_10CM,TSA,H2OSNO,TOTLITC,CWDC,SOIL1C,SOIL2C,SOIL3C,SOIL4C,WOOD_HARVESTC,TOTVEGC,NBP,GPP,AR,HR --plots_ocn None --nrows 4 --ncols 2 --results_dir ${results_dir} --regions glb,n,s --make_viewer True --start_yr 1985 --end_yr 1995` |
| Classic PDF for both original & component | `plots_original`, `plots_atm`, `plots_lnd` are set; `make_viewer` is not. | `zi-global-time-series --subsection ${subsection} --use_ocn True --input /lcrc/group/e3sm2/ac.wlin//E3SMv3/v3.LR.historical_0051 --input_subdir archive/ocn/hist --moc_file mocTimeSeries_1985-1995.nc --case_dir /lcrc/group/e3sm/ac.forsyth2/zppy_weekly_comprehensive_v3_output/test_issue-23-rebased-20250903/v3.LR.historical_0051 --experiment_name v3.LR.historical_0051 --figstr v3.LR.historical_0051 --color Blue --ts_num_years 5 --plots_original net_toa_flux_restom,global_surface_air_temperature,toa_radiation,net_atm_energy_imbalance,change_ohc,max_moc,change_sea_level,net_atm_water_imbalance --plots_atm TREFHT --plots_ice None --plots_lnd FSH,RH2M,LAISHA,LAISUN,QINTR,QOVER,QRUNOFF,QSOIL,QVEGE,QVEGT,SOILWATER_10CM,TSA,H2OSNO,TOTLITC,CWDC,SOIL1C,SOIL2C,SOIL3C,SOIL4C,WOOD_HARVESTC,TOTVEGC,NBP,GPP,AR,HR --plots_ocn None --nrows 4 --ncols 2 --results_dir ${results_dir} --regions glb,n,s --make_viewer False --start_yr 1985 --end_yr 1995` |
| Classic PDF for just original | `plots_original` is set; `plots_atm`, `plots_lnd`, `make_viewer` are not. | `zi-global-time-series --subsection ${subsection} --use_ocn True --input /lcrc/group/e3sm2/ac.wlin//E3SMv3/v3.LR.historical_0051 --input_subdir archive/ocn/hist --moc_file mocTimeSeries_1985-1995.nc --case_dir /lcrc/group/e3sm/ac.forsyth2/zppy_weekly_comprehensive_v3_output/test_issue-23-rebased-20250903/v3.LR.historical_0051 --experiment_name v3.LR.historical_0051 --figstr v3.LR.historical_0051 --color Blue --ts_num_years 5 --plots_original net_toa_flux_restom,global_surface_air_temperature,toa_radiation,net_atm_energy_imbalance,change_ohc,max_moc,change_sea_level,net_atm_water_imbalance --plots_atm None --plots_ice None --plots_lnd None --plots_ocn None --nrows 4 --ncols 2 --results_dir ${results_dir} --regions glb,n,s --make_viewer False --start_yr 1985 --end_yr 1995`|
| Classic PDF for just component | `plots_atm`, `plots_lnd`, are set; `plots_original`, `make_viewer` are not. | `zi-global-time-series --subsection ${subsection} --use_ocn False --input /lcrc/group/e3sm2/ac.wlin//E3SMv3/v3.LR.historical_0051 --input_subdir archive/ocn/hist --moc_file None --case_dir /lcrc/group/e3sm/ac.forsyth2/zppy_weekly_comprehensive_v3_output/test_issue-23-rebased-20250903/v3.LR.historical_0051 --experiment_name v3.LR.historical_0051 --figstr v3.LR.historical_0051 --color Blue --ts_num_years 5 --plots_original None --plots_atm TREFHT --plots_ice None --plots_lnd FSH,RH2M,LAISHA,LAISUN,QINTR,QOVER,QRUNOFF,QSOIL,QVEGE,QVEGT,SOILWATER_10CM,TSA,H2OSNO,TOTLITC,CWDC,SOIL1C,SOIL2C,SOIL3C,SOIL4C,WOOD_HARVESTC,TOTVEGC,NBP,GPP,AR,HR --plots_ocn None --nrows 4 --ncols 2 --results_dir ${results_dir} --regions glb,n,s --make_viewer False --start_yr 1985 --end_yr 1995` |
| Classic PDF for just original, atm only | `plots_original` is set to only non-ocn plots; `plots_atm`, `plots_lnd`, `make_viewer` are not set. | `zi-global-time-series --subsection ${subsection} --use_ocn False --input /lcrc/group/e3sm2/ac.wlin//E3SMv3/v3.LR.historical_0051 --input_subdir archive/ocn/hist --moc_file None --case_dir /lcrc/group/e3sm/ac.forsyth2/zppy_weekly_comprehensive_v3_output/test_issue-23-rebased-20250903/v3.LR.historical_0051 --experiment_name v3.LR.historical_0051 --figstr v3.LR.historical_0051 --color Blue --ts_num_years 5 --plots_original net_toa_flux_restom,global_surface_air_temperature,toa_radiation,net_atm_energy_imbalance,net_atm_water_imbalance --plots_atm None --plots_ice None --plots_lnd None --plots_ocn None --nrows 4 --ncols 2 --results_dir ${results_dir} --regions glb,n,s --make_viewer False --start_yr 1985 --end_yr 1995`|
| Viewer for all land variables | `plots_lnd` is set to "all"; `make_viewer` is set; `plots_original`, `plots_atm` are not set. | `zi-global-time-series --subsection ${subsection} --use_ocn False --input /lcrc/group/e3sm2/ac.wlin//E3SMv3/v3.LR.historical_0051 --input_subdir archive/ocn/hist --moc_file None --case_dir /lcrc/group/e3sm/ac.forsyth2/zppy_weekly_comprehensive_v3_output/test_issue-23-rebased-20250903/v3.LR.historical_0051 --experiment_name v3.LR.historical_0051 --figstr v3.LR.historical_0051 --color Blue --ts_num_years 5 --plots_original None --plots_atm None --plots_ice None --plots_lnd all --plots_ocn None --nrows 4 --ncols 2 --results_dir ${results_dir} --regions glb,n,s --make_viewer True --start_yr 1985 --end_yr 1995`|

Note that `subsection` is irrelevant if you're calling `zppy-interfaces` directly. Perhaps use `zppy_cfg_subsection_placeholder`

> I'm wondering how to modify this to make the original plots work? 

Just get your `sys.argv` to match the desired call listed above. 

Notes/caveats:
- You'd probably want to run without the ocean plots, unless you already have the necessary files generated from a previous MPAS-Analysis run.
- Recall that in Viewer mode, only the original plot PNGs show up in the Viewer. The PDFs look like a broken link, but will download if you click on "download"

forsyth2 commented:
Also, I think I have the rebase sorted out (again, just logger import changes). I'm running a test to make sure it works ok and then I will push to this branch.

forsyth2 commented:
Ok, got the rebase taken care of. I've updated this branch; nothing should change (except for logs not printing multiple times). I also rebased the zppy branch, but that didn't have any conflicts to worry about.

<details>

<summary> Rebase details </summary>

```bash
git checkout -b issue-23-gts-split-rebased-20250903
git rebase -i f0d61bca47189dae1038469eb3d83cd92701e20b
# We'll keep 2 commits: that will make the rebase a little easier while still keeping the redesign distinct.
# 
# pick 572f077 Split coupled_global
# f 8dc1578 Get tests passing
# f 4e5b93c Address comments
# pick 1ca04de Redesign for backwards compatibility
git fetch upstream main
git rebase upstream/main
git status
	# deleted by them: zppy_interfaces/global_time_series/coupled_global.py
	# deleted by them: zppy_interfaces/global_time_series/coupled_global_dataset_wrapper.py
	# both modified:   zppy_interfaces/global_time_series/coupled_global_plotting.py
rm zppy_interfaces/global_time_series/coupled_global.py zppy_interfaces/global_time_series/coupled_global_dataset_wrapper.py
git status
	# deleted by them: zppy_interfaces/global_time_series/coupled_global.py
	# deleted by them: zppy_interfaces/global_time_series/coupled_global_dataset_wrapper.py
	# both modified:   zppy_interfaces/global_time_series/coupled_global_plotting.py
git grep -n "<<<<<<<"
# zppy_interfaces/global_time_series/coupled_global_plotting.py:8:<<<<<<< HEAD
#
# Edit file
git grep -n "<<<<<<<"
# No further results
git add zppy_interfaces/global_time_series/coupled_global.py zppy_interfaces/global_time_series/coupled_global_dataset_wrapper.py zppy_interfaces/global_time_series/coupled_global_plotting.py
git rebase --continue
git status
	# both modified:   zppy_interfaces/global_time_series/coupled_global/mix_viewer_component.py
	# both modified:   zppy_interfaces/global_time_series/create_ocean_ts.py
git grep -n "<<<<<<<"
# zppy_interfaces/global_time_series/coupled_global/mix_viewer_component.py:20:<<<<<<< HEAD:zppy_interfaces/global_time_series/viewer/coupled_global_viewer.py
# zppy_interfaces/global_time_series/create_ocean_ts.py:12:<<<<<<< HEAD:zppy_interfaces/global_time_series/classic/ocean_month.py
#
# Edit files
git grep -n "<<<<<<<"
# No further results
git add zppy_interfaces/global_time_series/coupled_global/mix_viewer_component.py zppy_interfaces/global_time_series/create_ocean_ts.py
git rebase --continue
git status
# nothing to commit

pre-commit run --all-files
# Fails
# Add child logger everywhere
pre-commit run --all-files
# Passes now
git add -A
git commit -m "Update logger imports post-rebase"

git status
# Nothing to commit
git checkout issue-23-gts-split
git branch -m issue-23-gts-split-backup-20250904
git checkout issue-23-gts-split-rebased-20250903
git log
# Update logger imports post-rebase
# Redesign for backwards compatibility
# Split coupled_global
# Add missing packages to dependencies (#32)
git branch -m issue-23-gts-split
git push -f upstream issue-23-gts-split
```

</details>

forsyth2 commented:
> Do we actually need subsection here?
> Note that `subsection` is irrelevant if you're calling `zppy-interfaces` directly. Perhaps use `zppy_cfg_subsection_placeholder`

It's not necessary but it's not set up as an optional parameter, so I would just go with the placeholder string for it.

> The zppy-interfaces inherited zppy's output parameter as case_dir, but zppy's output directory structure contains both existing processed data (to read from) and new result locations (to write to). In this case someone can't testing with the input data created by someone else. I'm trying to write out data (moc file) to result_dir, to prevent this problem.

This problem exists on `main`. On the `main` branch`, [__main__.py](https://github.com/E3SM-Project/zppy-interfaces/blob/main/zppy_interfaces/global_time_series/__main__.py) has the following comments regarding modification of directories:
```
# NOTE: MODIFIES THE CASE DIRECTORY (parameters.case_dir) post subdirectory
# Creates the directory post/ocn

# NOTE: MODIFIES THE CASE DIRECTORY (parameters.case_dir) post subdirectory
# Modifies post/ocn (which we just created in the first place)

# NOTE: MODIFIES THE CASE DIRECTORY (parameters.case_dir) post subdirectory
# Copies files to post/ocn (which we just created in the first place)

# NOTE: PRODUCES OUTPUT IN THE CURRENT DIRECTORY (not necessarily the case directory)
# Creates the directory parameters.results_dir
```

That is, a number of operations are done to modify `{case_dir}/post/ocn` and later operations may be expecting data to be located there. I'm inclined to say this issue may be out of scope of this PR, but if it's a simple fix, then perhaps it would be ok.

forsyth2 commented:
> Could you help me to fix the unit test error which was caused by the change of directory path?

Done with https://github.com/E3SM-Project/zppy-interfaces/pull/26/commits/8a0ea688d231a3d2e2de0e8592e5b70469fe080c

> Let me know if you notice any problems with my change.

I did a high-level visual inspection of your 5 commits. I'll also do a run with `zppy` to see if any issues pop up.

> I will move to test zppy next.

Great, thanks!

forsyth2 commented:
>  I will move to test zppy next.
Note you'll need to remove update `zppy/templates/global_time_series.bash` accordingly, or else the call will fail due to unknown parameter.

```
git diff --staged zppy/templates/global_time_series.bash
```
```
-{% if subsection != None %}
-subsection={{ subsection }}
-{%- else %}
-subsection="no_subsection"
-{%- endif %}
-
 results_dir={{ prefix }}_results
-zi-global-time-series --subsection ${subsection} --use_ocn {{ use_ocn }} --input {{ input }} --input_subdir {{ input_subdir }} --moc_file {{ moc_file }} --case_dir {{ output }} --experiment_name {{ experiment_name }} --figstr {{ figstr }} --color {{ color }} --ts_num_years {{ ts_num_years }} --plots_original {{ plots_original }} --plots_atm {{ plots_atm }} --plots_ice {{ plots_ice }} --plots_lnd {{ plots_lnd }} --plots_ocn {{ plots_ocn }} --nrows {{ nrows }} --ncols {{ ncols }} --results_dir ${results_dir} --regions {{ regions }} --make_viewer {{ make_viewer }} --start_yr {{ year1 }} --end_yr {{ year2 }}
+zi-global-time-series --use_ocn {{ use_ocn }} --input {{ input }} --input_subdir {{ input_subdir }} --moc_file {{ moc_file }} --case_dir {{ output }} --experiment_name {{ experiment_name }} --figstr {{ figstr }} --color {{ color }} --ts_num_years {{ ts_num_years }} --plots_original {{ plots_original }} --plots_atm {{ plots_atm }} --plots_ice {{ plots_ice }} --plots_lnd {{ plots_lnd }} --plots_ocn {{ plots_ocn }} --nrows {{ nrows }} --ncols {{ ncols }} --results_dir ${results_dir} --regions {{ regions }} --make_viewer {{ make_viewer }} --start_yr {{ year1 }} --end_yr {{ year2 }}
```

chengzhuzhang commented:
When testing with 350+ land variables, multiprocessing was causing deadlocks and memory issues. I switched back to use sequential processing provides better reliability and memory management. The refactored code with computation and plotting simplified and combined already has good performance even with sequential processing. For all land variables, it take less than **14 mins ** to create all plots. The test is based on 10 years of global datasets, we should not expect memory issue with longer datasets with this setting.

forsyth2 commented:
> Could you re-install and try if this fix the issue?

## Testing process

<details>

<summary> Expand </summary>

```bash
cd ~/ez/zppy-interfaces
git status
# Branch: branch add-pcmdi
# No uncommitted changes
git checkout issue-23-gts-split
git log
# Last commit (9/18): fix pre-committing
# Bad, missing latest commit from https://github.com/E3SM-Project/zppy-interfaces/pull/26/commits
# Good, based on top of this commit: Add missing packages to dependencies (#32)
# Which matches https://github.com/E3SM-Project/zppy-interfaces/commits/main/
# It only has 15 (rather than 17) commits on top though.
# That's because 1 is a merge commit and 1 is the commit we've haven't included yet.
# Also, the commit hashs don't match https://github.com/E3SM-Project/zppy-interfaces/pull/26/commits,
# presumably because we did a rebase and didn't push the changes to the branch.
# So, let's update this branch accordingly.
# First, make a backup
git checkout -b issue-23-gts-split-backup20250930
git checkout issue-23-gts-split
git fetch upstream
git reset --hard upstream/issue-23-gts-split
git log
# Good, latest commit: fix directory generation for component only config
# Good, commit hashs match https://github.com/E3SM-Project/zppy-interfaces/pull/26/commits
# Good, 17 commits on top of: Add missing packages to dependencies (#32)
# Now, to rebase off the latest `main`
git fetch upstream
git rebase upstream/main
git log
# Good, 16 commits on top of: Add missing packages to dependencies (#32)
# (17 commits minus the now unnecessary merge commit)
rm -rf build
conda clean --all --y # ~7 min to run
conda env create -f conda/dev.yml -n zi-gts-updates-20250930 # ~3 min to run
conda activate zi-gts-updates-20250930
pre-commit run --all-files
python -m pip install .
pytest tests/unit/global_time_series/test_*.py
# 10 passed in 27.02s

cd ~/ez/zppy
git status
# Branch: zppy_pcmdi_202505_rev
# Handle uncommitted changes:
git add -A
git commit -m "Testing" --no-verify
git checkout issue-23-gts-split
git log
# Last commit (9/19): Testing
# Let's reset the branch
git checkout -b issue-23-gts-split-backup20250930
git checkout issue-23-gts-split
git fetch upstream
git reset --hard upstream/issue-23-gts-split
# And rebase off the latest `main`:
git rebase upstream/main
# Handle conflicts (repeat for second commit)
git grep -n "<<<<<<<"
# tests/integration/utils.py:42:<<<<<<< HEAD
# Accept Current Change
git grep -n "<<<<<<<"
# No remaining conflicts
git add tests/integration/utils.py
git rebase --continue
git log
# Good, 3 commits on top of: Integration test fixes (#737)
rm -rf build
conda clean --all --y
conda env create -f conda/dev.yml -n zppy-gts-updates-20250930
conda activate zppy-gts-updates-20250930
pytest tests/test_*.py
# 25 passed in 0.92s

# Edit tests/integration/utils.py
# TEST_SPECIFICS: Dict[str, Any] = {
#     "diags_environment_commands": "source /gpfs/fs1/home/ac.forsyth2/miniforge3/etc/profile.d/conda.sh; conda activate e3sm-diags-main-20250926-rerun",
#     "global_time_series_environment_commands": "source /gpfs/fs1/home/ac.forsyth2/miniforge3/etc/profile.d/conda.sh; conda activate zi-gts-updates-20250930",
#     "cfgs_to_run": [
#         "weekly_bundles",
#         "weekly_comprehensive_v2",
#         "weekly_comprehensive_v3",
#         "weekly_legacy_3.0.0_bundles",
#         "weekly_legacy_3.0.0_comprehensive_v2",
#         "weekly_legacy_3.0.0_comprehensive_v3",
#     ],
#     "tasks_to_run": ["global_time_series"],
#     "unique_id": "test_gts_updates_20250930",
# }
python tests/integration/utils.py
pre-commit run --all-files
python -m pip install .
zppy -c tests/integration/generated/test_weekly_bundles_chrysalis.cfg
zppy -c tests/integration/generated/test_weekly_comprehensive_v2_chrysalis.cfg
zppy -c tests/integration/generated/test_weekly_comprehensive_v3_chrysalis.cfg
zppy -c tests/integration/generated/test_weekly_legacy_3.0.0_bundles_chrysalis.cfg
zppy -c tests/integration/generated/test_weekly_legacy_3.0.0_comprehensive_v2_chrysalis.cfg
zppy -c tests/integration/generated/test_weekly_legacy_3.0.0_comprehensive_v3_chrysalis.cfg

# Picking up Oct. 1 ###########################################################
# Check on bundles status
cd /lcrc/group/e3sm/ac.forsyth2/zppy_weekly_bundles_output/test_gts_updates_20250930/v3.LR.historical_0051/post/scripts
grep -v "OK" *status
# Good, no errors
cd /lcrc/group/e3sm/ac.forsyth2/zppy_weekly_legacy_3.0.0_bundles_output/test_gts_updates_20250930/v3.LR.historical_0051/post/scripts
grep -v "OK" *status
# Good, no errors
# Now, run bundles part 2
cd ~/ez/zppy
git status
# No uncommitted changes
# Branch: zppy_pcmdi_202505_rev
git checkout issue-23-gts-split
conda activate zppy-gts-updates-20250930
zppy -c tests/integration/generated/test_weekly_bundles_chrysalis.cfg
zppy -c tests/integration/generated/test_weekly_legacy_3.0.0_bundles_chrysalis.cfg
# Actually, nothing new to run when we're only running global_time_series

### v2  ###
cd /lcrc/group/e3sm/ac.forsyth2/zppy_weekly_comprehensive_v2_output/test_gts_updates_20250930/v2.LR.historical_0201/post/scripts
grep -v "OK" *status
# Good, no errors
cd /lcrc/group/e3sm/ac.forsyth2/zppy_weekly_legacy_3.0.0_comprehensive_v2_output/test_gts_updates_20250930/v2.LR.historical_0201/post/scripts
grep -v "OK" *status
# Good, no errors
### v3 ###
cd /lcrc/group/e3sm/ac.forsyth2/zppy_weekly_comprehensive_v3_output/test_gts_updates_20250930/v3.LR.historical_0051/post/scripts
grep -v "OK" *status
# Good, no errors
# This is an improvement from the last run!
cd /lcrc/group/e3sm/ac.forsyth2/zppy_weekly_legacy_3.0.0_comprehensive_v3_output/test_gts_updates_20250930/v3.LR.historical_0051/post/scripts
grep -v "OK" *status
# Good, no errors

cd ~/ez/zppy
ls tests/integration/test_*.py
git status
# Branch: issue-23-gts-split
pytest tests/integration/test_bash_generation.py
# rm: cannot remove 'test_bash_generation_output/post/scripts/provenance*': No such file or directory
# 1 failed in 1.12s
# This appears to be because a `rm` in a test fix needed to be `rm -rf`
# Add the `-rf` to those commands.
# Retest:
pytest tests/integration/test_bash_generation.py
# 1 failed in 2.11s
# Diffs appear to be expected, based on this PR's changes to zppy/templates/global_time_series.bash
pytest tests/integration/test_campaign.py
# 6 passed in 3.34s
pytest tests/integration/test_defaults.py
# 1 passed in 0.47s
pytest tests/integration/test_last_year.py
# 1 passed in 0.49s
pytest tests/integration/test_bundles.py
# 2 failed in 0.51s
# Failure is because we didn't run the entire cfg and 
# this test isn't customized like test_images.py is.
pytest tests/integration/test_images.py # 9:49
# 1 failed in 15.00s

# Add fix to test_images.py
pytest tests/integration/test_images.py
# FAILED tests/integration/test_images.py::test_images - assert 6 == 3
# 1 failed in 9.31s
cat test_images_summary.md
```

</details>

## Complete results

<details>

<summary> Expand </summary>

Diff subdir is where to find the lists of missing/mismatched images, the image diff grid, and the individual diffs.
Note image diff grids can not yet be constructed automatically.
| Test name | Total images | Correct images | Missing images | Mismatched images | Diff subdir | 
| --- | --- | --- | --- | --- | --- | 
| comprehensive_v2_global_time_series | 6 | 6 | 0 | 0 | /lcrc/group/e3sm/public_html/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v2_www/test_gts_updates_20250930/v2.LR.historical_0201/image_check_failures_comprehensive_v2/global_time_series | 
| comprehensive_v3_global_time_series | 15 | 0 | 15 | 0 | /lcrc/group/e3sm/public_html/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v3_www/test_gts_updates_20250930/v3.LR.historical_0051/image_check_failures_comprehensive_v3/global_time_series | 
| bundles_global_time_series | 3 | 3 | 0 | 0 | /lcrc/group/e3sm/public_html/diagnostic_output/ac.forsyth2/zppy_weekly_bundles_www/test_gts_updates_20250930/v3.LR.historical_0051/image_check_failures_bundles/global_time_series | 
| legacy_3.0.0_comprehensive_v2_global_time_series | 6 | 6 | 0 | 0 | /lcrc/group/e3sm/public_html/diagnostic_output/ac.forsyth2/zppy_weekly_legacy_3.0.0_comprehensive_v2_www/test_gts_updates_20250930/v2.LR.historical_0201/image_check_failures_legacy_3.0.0_comprehensive_v2/global_time_series | 
| legacy_3.0.0_comprehensive_v3_global_time_series | 15 | 15 | 0 | 0 | /lcrc/group/e3sm/public_html/diagnostic_output/ac.forsyth2/zppy_weekly_legacy_3.0.0_comprehensive_v3_www/test_gts_updates_20250930/v3.LR.historical_0051/image_check_failures_legacy_3.0.0_comprehensive_v3/global_time_series | 
| legacy_3.0.0_bundles_global_time_series | 3 | 3 | 0 | 0 | /lcrc/group/e3sm/public_html/diagnostic_output/ac.forsyth2/zppy_weekly_legacy_3.0.0_bundles_www/test_gts_updates_20250930/v3.LR.historical_0051/image_check_failures_legacy_3.0.0_bundles/global_time_series | 

</details>

## Concise results

| Test name | Total images | Correct images | Missing images | Mismatched images | Diff subdir | 
| --- | --- | --- | --- | --- | --- | 
| comprehensive_v3_global_time_series | 15 | 0 | 15 | 0 | /lcrc/group/e3sm/public_html/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v3_www/test_gts_updates_20250930/v3.LR.historical_0051/image_check_failures_comprehensive_v3/global_time_series | 

And those failures are because we changed what subtasks run for the v3 test, so all good.

forsyth2 commented:
Not yet, I need to push rebased versions to avoid the merge conflict. Doing that now

chengzhuzhang commented:
@forsyth2 I notice that the time-series plot always has x axis (year) longer than the data and having the average value pushed away to the right ([example](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v3_www/test_gts_updates_20250930/v3.LR.historical_0051/global_time_series/global_time_series_classic_pdf_original_1985-1995_results/v3.LR.historical_0051_glb_original.pdf), is this expected? I think I saw this behavior since working on this branch...

chengzhuzhang commented:
The atm variables are now showing up. I suppose it is another configuration issue, as the input may not be available for testing?

forsyth2 commented:
> The atm variables are now showing up

No, unfortunately, this appears to be a new issue. 

I can confirm the links aren't showing up in https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v3_www/test_gts_updates_20250930/v3.LR.historical_0051/global_time_series/global_time_series_viewer_both_1985-1995_results/table_atm/index.html.

```bash
cd /lcrc/group/e3sm/public_html/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v3_www/test_gts_updates_20250930/v3.LR.historical_005
cat provenance.20250930_220814_975564.cfg
# [ts] > [[ atm_monthly_glb ]] doesn't specify vars so we must be using the default vars
cd /lcrc/group/e3sm/ac.forsyth2/zppy_weekly_comprehensive_v3_output/test_gts_updates_20250930/v3.LR.historical_0051/post/scripts
grep -n vars ts_atm_monthly_glb_1985-1989-0005.settings
# 15:  'extra_vars': '',
# 47:  'vars': 'FSNTOA,FLUT,FSNT,FLNT,FSNS,FLNS,SHFLX,QFLX,TAUX,TAUY,PRECC,PRECL,PRECSC,PRECSL,TS,TREFHT,CLDTOT,CLDHGH,CLDMED,CLDLOW,U',
```
We can see that `TREFHT` is in fact present.
```bash
grep -n TREFHT *.o*
# This isn't particularly illuminating.
grep -in error *.o*
Also not particularly illuminating.
```

forsyth2 commented:
The atmosphere omission also only occurs for viewers. They show up on the PDFs -- https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v3_www/test_gts_updates_20250930/v3.LR.historical_0051/global_time_series/global_time_series_classic_pdf_both_1985-1995_results/v3.LR.historical_0051_glb_atm.pdf

chengzhuzhang commented:
I tested this new fix and the atm variable show up. It seems to be a bug that prevent single variable to be plotted. Though I didn't know why the single variable plot with viewer did show up in my early test. @forsyth2, could you test again? Thanks.

chengzhuzhang commented:
Great news. I'm not very clear about the cases you tested: what is the difference between original only and Original 8 plots? Otherwise, I think this PR is good to merge.

forsyth2 commented:
Those last two rows are sort of extra add-on cases. The top 3 rows are the basics.

chengzhuzhang commented:
OH, I mis interpreted the table. Thanks for clarifying. If N/A just means the case was not tested, then we are okay.
